### PR TITLE
Finalize toolchain spec properties when resolving a toolchain

### PR DESCRIPTION
**Pull request by alllex**
Even though implementations of the `JavaToolchainSpec` interface all belong to internal API, we still must make sure that users or plugins that wire some changes to the specs are not corrupting the toolchain cache. Additionally, even though the toolchain API discourages users to be able to obtain instances of toolchain specs used for toolchain resolution, this could potentially change in the future.

However, there is at least one place in the API, where the users are able to have a reference to an actual spec instance and modify it at any point:
https://github.com/gradle/gradle/blob/ba32027bf0656be5c8a71e6281939ff410a9cf1a/subprojects/plugins/src/main/java/org/gradle/api/plugins/JavaPluginExtension.java#L164

One of the places where this can happen is during the toolchain resolution when we query the toolchain spec key multiple times. Any access to the spec property values durnig the resolution is also a potential subject of the problem.

This PR makes sure that once a toolchain spec is used for toolchain resolution, values of all it's properties are finalized.

### CODE REVIEW

**Comment by alllex** [comment]
@bot-gradle test qfl

**Comment by alllex** [comment]
@bot-gradle test this

**Review by jbartok — APPROVED** [review]

**Comment by alllex** [comment]
@bot-gradle test qfl

**Comment by alllex** [comment]
@bot-gradle test this

**Comment by alllex** [comment]
@bot-gradle test qfl

File: subprojects/platform-jvm/src/integTest/groovy/org/gradle/jvm/toolchain/JavaToolchainIntegrationTest.groovy [hunk 1]
```diff
@@ -101,6 +115,63 @@ class JavaToolchainIntegrationTest extends AbstractIntegrationSpec {
...
+        buildScript """
+            apply plugin: "java"
+
+            java {
+                toolchain {
+                    languageVersion = JavaLanguageVersion.of(${jdkMetadata1.languageVersion.majorVersion})
+                }
+            }
+
+            compileJava {
```
Comment (wolfs) [inline subprojects/platform-jvm/src/integTest/groovy/org/gradle/jvm/toolchain/JavaToolchainIntegrationTest.groovy:135]: Would it be better to use a custom task here to show when the first querying happens?

File: subprojects/platform-jvm/src/main/java/org/gradle/jvm/toolchain/internal/JavaToolchainQueryService.java [hunk 2]
```diff
@@ -80,38 +90,22 @@ ProviderInternal<JavaToolchain> findMatchingToolchain(JavaToolchainSpec filter)
...
+            return null;
+        }
 
-            synchronized (matchingToolchains) {
-                if (matchingToolchains.containsKey(filterInternal.toKey())) {
-                    return handleMatchingToolchainCached(filterInternal);
-                } else {
-                    return handleMatchingToolchainUnknown(filterInternal);
-                }
+        Object resolutionResult = matchingToolchains.computeIfAbsent(filterInternal.toKey(), key -> {
```
Comment (wolfs) [inline subprojects/platform-jvm/src/main/java/org/gradle/jvm/toolchain/internal/JavaToolchainQueryService.java:97]: In order to drop the synchronization here, doesn't `matchingToolchains` need to be a ConcurrentHashMap` and not just a `HashMap`?

Comment (alllex) [reply subprojects/platform-jvm/src/main/java/org/gradle/jvm/toolchain/internal/JavaToolchainQueryService.java:97]: Good catch! I was sure it was a concurrent map before, but it seems that has changed after the download SPI has been merged.

File: subprojects/platform-jvm/src/integTest/groovy/org/gradle/jvm/toolchain/JavaToolchainIntegrationTest.groovy (hunk 1 above)
Comment (alllex) [reply subprojects/platform-jvm/src/integTest/groovy/org/gradle/jvm/toolchain/JavaToolchainIntegrationTest.groovy]: You are right, it's enough to explicitly resolve the java extension toolchain during the configuration time.

**Comment by alllex** [comment]
@bot-gradle test qfl

**Review by wolfs — APPROVED** [review]
LGTM!

**Comment by alllex** [comment]
@bot-gradle test and merge

**Pull request merged**